sklearn/metric_learning/nca.py

```diff
@@ -0,0 +1,280 @@
+from ..base import BaseEstimator, TransformerMixin
+from ..utils.validation import check_is_fitted
+
+import numpy as np
+import scipy as sp
+import scipy.misc
+import scipy.optimize
+import time
+
+import sys
+
+
+def nca_vectorized_oracle(X, y, n_components, loss, threshold=0.0):
+    dx = X[:, np.newaxis, :] - X[np.newaxis, :, :]  # n_samples x n_samples x n_features
+    outer = dx[:, :, np.newaxis, :] * dx[:, :, :, np.newaxis]  # n_samples x n_samples x n_features x n_features
+
+    n_samples, n_features = X.shape
+
+    def oracle(L):
```

Comment:
I think the nested function design could be avoided by instead having top-level `nca_vectorized_loss_grad` and `nca_semivectorized_loss_grad`, and then in `optimize_nca` doing:
```
if memory_permits:
    loss, grad = nca_vectorized_loss_grad(...)
else:
    loss, grad = nca_semivectorized_loss_grad(...)
```
unless I'm missing something here. The concern is that nested functions can't be pickled.